### DISCUSSION TITLE
Support renovation of HelmReleases without explicit namespace

### DISCUSSION THREAD

**Discussion by GfxKai** [renovatebot/renovate#20037] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

My team have opted to specifically avoid defining namespaces on flux-controlled resources. By grouping resources into folders and assigning them to a namespace via their parent kustomization, we can better organise our deployments and make refactoring easier and less error-prone.

Unfortunately, as stated in the [docs](https://docs.renovatebot.com/modules/manager/flux/#helmrelease-support):
> In addition, for the flux manager to properly link HelmRelease and HelmRepository resources, both of the following conditions must be met:
> 1. The HelmRelease resource must either have its metadata.namespace property set or its spec.chart.spec.sourceRef.namespace property set
> 2. The referenced HelmRepository resource must have its metadata.namespace property set
> 
> Namespaces will not be inferred from the context (e.g. from the parent Kustomization).

Our current workaround for this is to add a `renovate-placeholder` namespace to every `HelmRepository` and `HelmRelease` resource, which gets overridden by the parent kustomization during reconciliation. Whilst this works, it's messy and kinda defeats the point of using the kustomization-based approach.

I had a search but couldn't find any documentation explaining _why_ this decision has been made, so thought I'd ask here. Happy to discuss and potentially make a PR if possible—would be great to find a solution!

**Comment by viceice** [comment]
this is to map the helm release to a corresponding helm repository. renovate otherwise don't know which repo to use for a helm release.

**Reply by GfxKai** [reply]
One potential solution could be to find the root `kustomization.yaml` in the repo (or add this path as a configuration option).

Renovate could then run `kustomize build  -o resources.yaml`, which applies all kustomizations to generate a full set of all flux-controlled resources in a single YAML file, with all namespaces, configMapGenerators etc. applied.

This would allow Renovate to match releases to repos, without requiring explicit placeholder namespaces on all resources.

**Reply by rarkins** [reply]
Renovate does not call package managers during the lookup phase as it's too compute intensive.

**Reply by viceice** [reply]
we then also loose the original file names where the update needs to be applied

**Reply by GfxKai** [reply]
Understood. From my tests, running `kustomize build` is pretty lightweight but I can understand the reluctance.

How about building a "namespace-resource" tree by iterating through from the root kustomization and referencing that during lookup if a namespace isn't found?

**Reply by rarkins** [reply]
Can it be determined just from the files in the repo, or does it require reaching out to the internet to resolve things?

**Reply by GfxKai** [reply]
Yep, can all be done offline. I'm hoping it shouldn't be too involved: we can find all `kustomization.yaml` resources in the repo, check them for a `namespace` property and, if found, assign each listed resource (matching `flux.fileMatch`) to that namespace.

Example `kustomization.yaml`:

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: example-namespace
resources:
  - resource1.yaml
  - resource2.yaml
  - nested/resource3.yaml
  - folder
```

**Reply by rarkins** [reply]
If it can be done offline and without too complicated logic, we should definition aim to support it. It would mean switching the `kustomize` extract logic to use [extractAllPackageFiles](https://github.com/renovatebot/renovate/blob/9d538a5210432024de2e329016bdd11175166fcd/lib/modules/manager/types.ts#L254-L257) approach instead of `extractPackageFile`. Doing so means the manager has the opportunity to "cross reference" all the files before returning them as an array, instead of parsing them and returning individually. If you can make a reproduction repo for this then please raise a feature request issue and link to it.

**Reply by GfxKai** [reply]
Sounds great! So the `kustomize` manager would be in charge of generating this namespace metadata during extraction, but then the `flux` manager can read the `kustomize` package files during its own lookup phase?

Will find some time on Monday to get a repro set up and make an issue. Thanks for taking the time to walk me through!

**Reply by rarkins** [reply]
Renovate has a separation between managers, which extract dependencies and current versions, and "datasources", which look up all available versions.
